### PR TITLE
Option for reducing allocations in the span pipeline

### PR DESCRIPTION
**Pull request by jkwatson**
This is a Draft proposal for one way to approach solving #1191 

@bogdandrutu let me know what you think.

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/trace/data/SpanData.java [hunk 1]
```diff
@@ -16,63 +16,47 @@
...
-/**
- * Immutable representation of all data collected by the {@link io.opentelemetry.trace.Span} class.
- *
- * @since 0.1.0
- */
-@Immutable
-@AutoValue
-public abstract class SpanData {
 
+public interface SpanData {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/data/SpanData.java:32]: Leave the immutable annotation?

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/data/SpanData.java:32]: yep. will add it back.

**Comment by jkwatson** [comment]
One option would be to return a wrapper of the RERS that implemented the SpanData interface. That's just one allocation (plus the necessary resolving of the links/events). That would be very easy to implement (and memoize in the RERS if we want to only generate it one time). WDYT?

**Comment by bogdandrutu** [comment]
Something about these lines. I think we should start with splitting the SD then deciding on the approach to do with RERS :). But I think we will go with that proposal unless we can think of a better option.

**Comment by jkwatson** [comment]
Ok, I'll put in a PR that just creates the SpanData interface along with the (badly named) SpanDataImpl and keeps the same implementation, then we can move from there.

**Comment by jkwatson** [comment]
closing this, as we're proceeding with this in separate PRs

**Pull request closed without merging**